### PR TITLE
Added test coverage for ProductService, ProductCategoryService, integrations and api

### PR DESCRIPTION
**Pull request by Agarwalvidu**
This PR includes all the rquirements till Week 5 along with unit tests and integration tests. Please review and suggest improvements.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -10,8 +10,8 @@
     "@types/node": "^22.13.4",
     "@types/react": "^18.0.0",
     "@types/react-dom": "^18.0.0",
-    "react": "^19.0.0",
-    "react-dom": "^19.0.0",
+    "react": "^18.3.1",
+    "react-dom": "^18.3.1",
```
Comment (tavyandy97) [inline frontend/package.json:14]: Any specific reason we've downgraded to react@18.x?

File: frontend/product.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      // Step 4 - Render product tiles dynamically
+      data.products.forEach((product) => {
+        const tile = document.createElement("div");
+        tile.className = "product-tile";
+
+        tile.innerHTML = `
+            <div class="product-title">${product.name}</div>
+            <div class="product-price">₹${product.price}</div>
+            <p>${product.description || ""}</p>
+            <p class="product-brand">${product.brand || ""}</p>
```
Comment (tavyandy97) [inline frontend/product.js:18]: nitpick: Both of these tags can be conditionally added depending upon if the value of the variable inside them exists

File: frontend/product.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+          `;
+        container.appendChild(tile);
```
Comment (tavyandy97) [inline frontend/product.js:20]: Recommendation:
Instead of appending elements one by one, we can use `DocumentFragment` (`document.createDocumentFragment`), which will help us avoid incrementally altering the DOM.

Repainting the DOM is a relatively heavier operation and would be better if we did it only once.

File: frontend/product.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+document.addEventListener("DOMContentLoaded", () => {
+  const container = document.getElementById("product-container");
+
+  fetch("http://127.0.0.1:8000/api/mongo/products/")
+    .then((res) => res.json())
+    .then((data) => {
+      console.log("Fetched products:", data);
```
Comment (tavyandy97) [inline frontend/product.js:7]: Maybe we can add some validation checks on data? There might be some cases where the data might not have come or the API call might've failed.

File: backend/tests/seed_data.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+from products_mongo.models import Product, ProductCategory
+from products_mongo.test_connection import init_db
+
+init_db
```
Comment (tavyandy97) [inline backend/tests/seed_data.py]: Shouldn't this be a function call?

File: backend/tests/seed_data.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    product2 = Product(
+        name="T-Shirt",
+        description="Cotton t-shirt",
+        price=19.99,
+        brand="FashionWear",
+        quantity_in_warehouse=200,
+        category=category2
+    )
+    product2.save()
```
Comment (tavyandy97) [inline backend/tests/seed_data.py]: When writing integration tests, it's a better practice to add multiple entries just to simulate a better DB state

**Comment by tavyandy97** [comment]
For the `backend/tests/test_product_category_services.py` file, we don't have any test cases for invalid inputs or error cases. It'd be a good practice to add those as well

File: frontend/product.css [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    padding: 15px;
+    width: 200px;
+    transition: transform 0.3s ease;
+  }
+  
+  .product-tile:hover {
+    transform: scale(1.05);
+  }
+  
+  .product-title {
```
Comment (tavyandy97) [inline frontend/product.css:26]: The CSS for this selector is written two times.

File: frontend/product.css [hunk 8]
```diff
@@ -0,0 +1,49 @@
+body {
```
Comment (tavyandy97) [inline frontend/product.css:1]: Recommendation: 
Can consider using CSS variables like 
```
:root {
  --color-background: #f8f8f8;
  --color-price: #2e8b57;
  --color-shadow: rgba(0, 0, 0, 0.1);
}
```

**Comment by tavyandy97** [comment]
A couple of cases that would be good to cover in `backend/tests/test_product_services.py`:
- Validation tests for:
    - Price negative values
    - Invalid date ranges
    - Duplicate product names
    - Invalid category assignments
- Making a separate file for mock data

File: frontend/package.json (hunk 1 above)
Comment (Agarwalvidu) [reply frontend/package.json:14]: I downgraded to React 18.3.1 because I encountered compatibility issues with some dependency, I am unable to recall it. But I am actively testing with React 19 again, will upgrade.

File: frontend/product.js (hunk 2 above)
Comment (Agarwalvidu) [reply frontend/product.js:18]: Thank you for pointing that out! I have updated the code to conditionally render the tags only when there is content. I’ will definitely keep this in mind going forward.

File: frontend/product.js (hunk 3 above)
Comment (Agarwalvidu) [reply frontend/product.js:20]: Thanks a lot for the recommendation!
I was not aware of DocumentFragment before, this was a great opportunity to learn about it. I have refactored the code to use a fragment to optimize DOM updates and will definitely keep such performance considerations in mind for future implementations as well.

File: frontend/product.js (hunk 4 above)
Comment (Agarwalvidu) [reply frontend/product.js:7]: That totally makes sense, I have now added checks to handle unexpected or empty responses(like I have added a check if the length of products data is 0 then it would display "No products found." and other checks if there is some issue with the API call). Also kept the technical logs in the console for debugging.

File: backend/tests/seed_data.py (hunk 5 above)
Comment (Agarwalvidu) [reply backend/tests/seed_data.py]: Thank you for catching it, I missed the parenthesis. I have corrected it to init_db() so the DB initialization now runs properly.

File: backend/tests/seed_data.py (hunk 6 above)
Comment (Agarwalvidu) [reply backend/tests/seed_data.py]: I will surely add more product entries. Will keep this in mind for future test setups as well!

**Comment by Agarwalvidu** [comment]
I have ve now added test cases for missing fields, invalid IDs, and failure cases in update/delete operations in backend/tests/test_product_category_services.py. Will make sure to include such edge cases going forward too.
I have added validation tests for price negative values, invalid date ranges, invalid category and duplicate product names in backend/tests/test_product_services.py. I have seperate file for mock data as well.
Thank you pointing these things, I learned a lot while adding these validations and I will keep these in mind as well.

File: frontend/product.css (hunk 7 above)
Comment (Agarwalvidu) [reply frontend/product.css:26]: Thanks for pointing that out! I have cleaned up the duplicate CSS for .product-title.

File: frontend/product.css (hunk 8 above)
Comment (Agarwalvidu) [reply frontend/product.css:1]: Thank you for the suggestion! I have implemented the CSS variables to make the code more maintainable.

**Comment by Agarwalvidu** [comment]
@tavyandy97 Please review the recent changes! Thank you.

File: frontend/package.json (hunk 1 above)
Comment (tavyandy97) [reply frontend/package.json:14]: Sure thing, lemme know what the issue is whenever you run into it 👍🏻

**Comment by tavyandy97** [comment]
Seems like there's a redundant set of frontend files. Removing the old set of files would be a good idea 😄

File: backend/django_app/settings.py [hunk 9]
```diff
@@ -15,6 +15,9 @@
 
 # Build paths inside the project like this: BASE_DIR / 'subdir'.
 BASE_DIR = Path(__file__).resolve().parent.parent
```
Comment (tavyandy97) [inline backend/django_app/settings.py:17]: Seems like some duplicate lines have been added in this file

File: backend/tests/test_product_category_services.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        category = ProductCategoryService.create_category(mock_category)
+        self.assertEqual(category["title"], "Electronics")
+
+    @patch('products_mongo.product_category_repository.ProductCategoryRepository.create_category')
+    def test_create_category_with_invalid_input(self, mock_create_category):
+        """Test category creation with missing fields"""
+        mock_create_category.side_effect = ValueError("Title is required")
+
+        with self.assertRaises(ValueError):
+            ProductCategoryService.create_category({"description": "No title provided"})
```
Comment (tavyandy97) [inline backend/tests/test_product_category_services.py:22]: - If the `ValueError` is raised by the service due to input validation before calling the repository, then mocking the repository is irrelevant - it will never be called.
- If the repository is responsible for raising the ValueError, then mocking it makes sense. Now it's just testing the mocked behavior, not real logic.

File: frontend/product.js (hunk 4 above)
Comment (tavyandy97) [reply frontend/product.js:7]: A check needs to be placed on the status as well for non 2xx HTTP status codes

File: backend/tests/test_product_category_services.py (hunk 10 above)
Comment (tavyandy97) [reply backend/tests/test_product_category_services.py:22]: Same comment can be applied for other tests which have been added
